kube-scheduler crashes and restarts with panics in DefaultPreemption plugin

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->

#### What happened:

Kubernetes 1.19 (confirmed with 1.19.7 and 1.19.10)

kube-scheduler crashes and restarts with the following errors in `default_preemption.go`. 

1. **panic: runtime error: index out of range [0] with length 0** 

The problem code is line 389.  When `victims` is nil or `victims.Pods` is empty, the error happens.  If we skip `GetPodPriority ` when it's nil or empty, the error is gone. 
```
387         victims := nodesToVictims[node]
388         // highestPodPriority is the highest priority among the victims on this node.
389         highestPodPriority := podutil.GetPodPriority(victims.Pods[0])
```

```
panic: runtime error: index out of range [0] with length 0 [recovered]
panic: runtime error: index out of range [0] with length 0
goroutine 1406 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x109
panic(0x1c9fde0, 0xc00748cf48)
 /usr/local/Cellar/go/1.16.2/libexec/src/runtime/panic.go:965 +0x1b9
k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.pickOneNodeForPreemption(0xc007e3dbc0, 0x4, 0x4)
/Users/yuanchen/projects/aci/go/kubescheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:389 +0xd2c
k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.SelectCandidate(0xc0086e0840, 0x4, 0x4, 0xc008235800, 0x2068138)
/Users/yuanchen/projects/aci/go/kubescheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:331 +0x85
k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.(*DefaultPreemption).preempt(0xc000ac4b80, 0x20677d0, 0xc005849540, 0xc0085616e0, 0xc008235800, 0xc008561710, 0x2d80750, 0xc00820a8c0, 0x7f0937ef47d8, 0x30)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:135 +0x4bd
k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.(*DefaultPreemption).PostFilter(0xc000ac4b80, 0x20677d0, 0xc005849540, 0xc0085616e0, 0xc0071543a8, 0xc008561710, 0x0, 0x0)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:83 +0xf1
k8s.io/kubernetes/pkg/scheduler/framework/runtime.(*frameworkImpl).runPostFilterPlugin(0xc0002fd380, 0x20677d0, 0xc005849540, 0x7f0937a84040, 0xc000ac4b80, 0xc0085616e0, 0xc0071543a8, 0xc008561710, 0xc0084f3ab0, 0x5f5e100)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/runtime/framework.go:557 +0x87
```

2. **panic: runtime error: invalid memory address or nil poster dereference**

` 452   latestStartTime := util.GetEarliestPodStartTime(nodesToVictims[minNodes2[0]])`

The problem is `nodesToVictims[minNodes2[0]]` does not exist and returns nil sometimes. Simply skipping it won't solve the problem.  The scheduler will reach either line 456 or 341. 

```
456        klog.Errorf("earliestStartTime is nil for node %s. Should not reach here.", minNodes2[0])
```

`341     klog.Errorf("None candidate can be picked from %v.", candidates)`


```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x15706d7]
goroutine 1385 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x109
panic(0x1b448c0, 0x2d281b0)
/usr/local/Cellar/go/1.16.2/libexec/src/runtime/panic.go:965 +0x1b9
k8s.io/kubernetes/pkg/scheduler/util.GetEarliestPodStartTime(0x0, 0xc004d4d9e0)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/kubernetes/pkg/scheduler/util/utils.go:56 +0x37
k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.pickOneNodeForPreemption(0xc004d4d9e0, 0x4, 0x4)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:452+0x7a9
k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.SelectCandidate(0xc0066e5180, 0x4, 0x4, 0xc0062f5000, 0x2068138)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:331 +0x85
k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.(*DefaultPreemption).preempt(0xc000f9e880, 0x20677d0, 0xc004fa3340, 0xc0061b5290, 0xc0062f5000, 0xc0061b52c0, 0x2d80750, 0xc00603fc40, 0x7f6bcbc363f8, 0x30)
/Users/yuanchen/projects/aci/go/kube-scheduler/vendor/k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:135 +0x4bd
: k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption.(*DefaultPreemption).PostFilter(0xc000f9e880, 0x20677d0, 0xc004fa3340, 0xc0061b5290, 0xc002523ac8, 0xc0061b52c0, 0x0, 0x0)
```

#### What you expected to happen:

The scheduler works without failures.

#### How to reproduce it (as minimally and precisely as possible):

When there are preemptions in a cluster. 

#### Anything else we need to know?:

#### Environment:
- Kubernetes version (use `kubectl version`): v1.19.10 or v1.19.7
- Cloud provider or hardware configuration: bare metal 
- OS (e.g: `cat /etc/os-release`): centos 7 
- Kernel (e.g. `uname -a`):   Linux 5.4.77-7.el7pie #1 SMP Sat Nov 21 01:16:27 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

--- a/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go
+++ b/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go
@@ -334,7 +334,7 @@ func dryRunPreemption(ctx context.Context, fh framework.Handle,
 		nodeInfoCopy := potentialNodes[(int(offset)+i)%len(potentialNodes)].Clone()
 		stateCopy := state.Clone()
 		pods, numPDBViolations, status := selectVictimsOnNode(ctx, fh, stateCopy, pod, nodeInfoCopy, pdbs)
-		if status.IsSuccess() {
+		if status.IsSuccess() && len(pods) != 0 {
 			victims := extenderv1.Victims{
 				Pods:             pods,
 				NumPDBViolations: int64(numPDBViolations),
@@ -352,11 +352,14 @@ func dryRunPreemption(ctx context.Context, fh framework.Handle,
 			if nvcSize > 0 && nvcSize+vcSize >= numCandidates {
 				cancel()
 			}
-		} else {
-			statusesLock.Lock()
-			nodeStatuses[nodeInfoCopy.Node().Name] = status
-			statusesLock.Unlock()
+			return
 		}
+		if status.IsSuccess() && len(pods) == 0 {
+			status = framework.AsStatus(fmt.Errorf("expected at least one victim pod on node %q", nodeInfoCopy.Node().Name))
+		}
+		statusesLock.Lock()
+		nodeStatuses[nodeInfoCopy.Node().Name] = status
+		statusesLock.Unlock()
 	}
 	fh.Parallelizer().Until(parallelCtx, len(potentialNodes), checkNode)
 	return append(nonViolatingCandidates.get(), violatingCandidates.get()...), nodeStatuses
@@ -391,6 +394,18 @@ func CallExtenders(extenders []framework.Extender, pod *v1.Pod, nodeLister frame
 			}
 			return nil, framework.AsStatus(err)
 		}
+		// Check if the returned victims are valid.
+		for nodeName, victims := range nodeNameToVictims {
+			if victims == nil || len(victims.Pods) == 0 {
+				if extender.IsIgnorable() {
+					delete(nodeNameToVictims, nodeName)
+					klog.InfoS("Ignoring node without victims", "node", nodeName)
+					continue
+				}
+				return nil, framework.AsStatus(fmt.Errorf("expected at least one victim pod on node %q", nodeName))
+			}
+		}
+
 		// Replace victimsMap with new result after preemption. So the
 		// rest of extenders can continue use it as parameter.
 		victimsMap = nodeNameToVictims

--- a/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
+++ b/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
@@ -357,13 +357,9 @@ func TestDryRunPreemption(t *testing.T) {
 				st.MakePod().Name("p1").UID("p1").Node("node1").Priority(midPriority).Obj(),
 				st.MakePod().Name("p2").UID("p2").Node("node2").Priority(midPriority).Obj(),
 			},
-			expected: [][]Candidate{
-				{
-					&candidate{victims: &extenderv1.Victims{}, name: "node1"},
-					&candidate{victims: &extenderv1.Victims{}, name: "node2"},
-				},
-			},
-			expectedNumFilterCalled: []int32{4},
+			expected:                [][]Candidate{{}},
+			fakeFilterRC:            framework.Unschedulable,
+			expectedNumFilterCalled: []int32{2},
 		},
 		{
 			name: "a pod that fits on one node with no preemption",
@@ -379,12 +375,9 @@ func TestDryRunPreemption(t *testing.T) {
 				st.MakePod().Name("p1").UID("p1").Node("node1").Priority(midPriority).Obj(),
 				st.MakePod().Name("p2").UID("p2").Node("node2").Priority(midPriority).Obj(),
 			},
-			expected: [][]Candidate{
-				{
-					&candidate{victims: &extenderv1.Victims{}, name: "node1"},
-				},
-			},
-			expectedNumFilterCalled: []int32{3},
+			expected:                [][]Candidate{{}},
+			fakeFilterRC:            framework.Unschedulable,
+			expectedNumFilterCalled: []int32{2},
 		},
 		{
 			name: "a pod that fits on both nodes when lower priority pods are preempted",
@@ -1073,16 +1066,6 @@ func TestSelectBestCandidate(t *testing.T) {
 		expected       []string // any of the items is valid
 	}{
 		{
-			name:           "No node needs preemption",
-			registerPlugin: st.RegisterPluginAsExtensions(noderesources.FitName, noderesources.NewFit, "Filter", "PreFilter"),
-			nodeNames:      []string{"node1"},
-			pod:            st.MakePod().Name("p").UID("p").Priority(highPriority).Req(largeRes).Obj(),
-			pods: []*v1.Pod{
-				st.MakePod().Name("p1").UID("p1").Node("node1").Priority(midPriority).Req(smallRes).StartTime(epochTime).Obj(),
-			},
-			expected: []string{"node1"},
-		},
-		{
 			name:           "a pod that fits on both nodes when lower priority pods are preempted",
 			registerPlugin: st.RegisterPluginAsExtensions(noderesources.FitName, noderesources.NewFit, "Filter", "PreFilter"),
 			nodeNames:      []string{"node1", "node2"},
@@ -1263,6 +1246,9 @@ func TestSelectBestCandidate(t *testing.T) {
 			offset, numCandidates := pl.getOffsetAndNumCandidates(int32(len(nodeInfos)))
 			candidates, _ := dryRunPreemption(context.Background(), fwk, state, tt.pod, nodeInfos, nil, offset, numCandidates)
 			s := SelectCandidate(candidates)
+			if s == nil || len(s.Name()) == 0 {
+				return
+			}
 			found := false
 			for _, nodeName := range tt.expected {
 				if nodeName == s.Name() {

